forc-test/src/lib.rs

```diff
@@ -115,24 +117,117 @@ pub struct TestPrintOpts {
 
 /// The storage and the contract id (if a contract is being tested) for a test.
 #[derive(Debug)]
-struct TestSetup {
+enum TestSetup {
+    ContractSetup(ContractTestSetup),
+    NonContractSetup(vm::storage::MemoryStorage),
+}
+
+impl TestSetup {
+    /// Returns the storage for this test setup
+    fn storage(&self) -> &vm::storage::MemoryStorage {
+        match self {
+            TestSetup::ContractSetup(contract_setup) => &contract_setup.storage,
+            TestSetup::NonContractSetup(storage) => storage,
+        }
+    }
+
+    /// Produces an iterator yielding contract ids of contract dependencies for this test setup.
+    fn contract_dependency_ids(&self) -> impl Iterator<Item = &tx::ContractId> + '_ {
+        match self {
+            TestSetup::ContractSetup(contract_setup) => {
+                contract_setup.contract_dependency_ids.iter()
+            }
+            TestSetup::NonContractSetup(_) => [].iter(),
+        }
+    }
+
+    /// Return the root contract id if this is a contract setup.
+    fn root_contract_id(&self) -> Option<tx::ContractId> {
+        if let TestSetup::ContractSetup(contract_setup) = self {
+            Some(contract_setup.root_contract_id)
+        } else {
+            None
+        }
+    }
+
+    /// Produces an iterator yielding all contract ids required to be included in the transaction
+    /// for this test setup.
+    fn contract_ids(&self) -> impl Iterator<Item = tx::ContractId> + '_ {
+        self.contract_dependency_ids()
+            .cloned()
+            .chain(self.root_contract_id().into_iter())
+    }
+}
+
+/// The data collected to test a contract.
+#[derive(Debug)]
+struct ContractTestSetup {
     storage: vm::storage::MemoryStorage,
-    contract_id: Option<tx::ContractId>,
+    contract_dependency_ids: Vec<tx::ContractId>,
+    root_contract_id: tx::ContractId,
+}
+
+impl ContractToTest {
+    /// Deploy the contract dependencies and tests excluded version for this package.
+    fn deploy(&self) -> anyhow::Result<TestSetup> {
+        // Setup the interpreter for deployment.
+        let params = tx::ConsensusParameters::default();
+        let storage = vm::storage::MemoryStorage::default();
+        let mut interpreter =
+            vm::interpreter::Interpreter::with_storage(storage, params, GasCosts::default());
+
+        // Root contract is the contract that we are going to be running the tests of, after this
+        // deployment.
+        let (root_contract_id, root_contract_tx) =
+            deployment_transaction(&self.pkg, &self.without_tests_bytecode, params);
```

Comment:
It looks like we don't need these until the final `interpreter.transact` call - perhaps we should move this call to after the contract dependency transactions have executed?